### ISSUE TITLE
Consider rewriting edit/runes to use UTF-8

### ISSUE THREAD

**Issue by eaburns** [eaburns/T#134]
This would have the benefit of eliminating copying, requiring less memory, and eliminating the need for rune->byte conversions, and we could use io.Reader, io.Writer, and friends instead of Runes versions of these.

Checkout https://godoc.org/golang.org/x/exp/utf8string for inspiration.

**Comment by eaburns** [comment]
Also consider a piece table: https://www.cs.unm.edu/~crowley/papers/sds.pdf

**Comment by eaburns** [comment]
### Piece table
There is a single file called the Add File. The only thing you do to the Add File is append to its end.
Then there is the Piece Table. The Piece Table points to Pieces of the Add File, and logically defines the data in the buffer.

type Piece struct {
    // Offs is the offset of the start of this piece in the Add File.
    offs int64
    // Size is the size of the piece (in runes, or bytes, or something).
    size int64
}

type PieceTable []Piece

The Buffer is the data found by scanning forward along the Piece Table, concatenating the contents of each Piece. (Of course you never really do such a full scan, unless you are doing a Put.)

To insert into the buffer:
1. Append the inserted data to the Add Table.
2. Make a Piece pointing to the inserted data (offs and size).
3. Put the piece into the table:
  * If the insertion point is in the middle of a Piece, split that Piece into two and insert between them (This is not touching the Add File data at all, just the Piece structs and the PieceTable slice.)
  * If the insertion point is between two existing pieces, simply insert the Piece in the table between them.
  * (There is an optimization in the 'between two pieces' case. If the preceding Piece's offs is the very end of the Add File, i.e., just before where the inserted data was appended, just increment that piece's size. This would actually be a common case if someone is adding text after a cursor.)

To delete (in a bit less detail, but you get the point):
1. If the prefix of a piece is being deleted, increment its offset.
2. If the suffix of a piece is being deleted, decrement its size.
3. If the piece size drops to 0, remove it from the table.

#### Undo/Redo
Since the Add File is only appended, all of the old data is in there. Undo/Redo are trivial, because restoring an old version of the Piece Table restores the buffer to exactly that state. So, this can be achieved by keeping a stack of piece tables.

#### Copy on write
This is an optimization, but in the paper he makes it a main point of the data structure. The idea is that, because the only file system operation is Append, previous Pieces are, in a sense, immutable. This trivially allows for copy on write when loading a file:
Initialize the Piece Table with a single Piece pointing at the original file data instead of a chunk of the Add File
Subsequent operations all modify the Add File, but un-touched Pieces point to the original file data ­— no copies needed.

However, we probably don't want to do this. If the file is change out from under us, then our data structure gets messed up.

**Comment by eaburns** [comment]
Since this is a fairly invasive change my though is to break it into as many small pieces as possible:
 1. Add runes.Buffer2 — a piece table implementation of runes.Buffer, still using runes. Change edit.Buffer.runes to be a runes.Buffer2. Everything else remains the same.
 2. Add undo/redo support to runes.Buffer2. Change edit Undo and Redo to use this instead of the runes.Buffer undo and redo logs.
 3. Change Editor.pend to use the runes.Buffer2. If the pending log and the actual buffer use the same Add File, then copying pending changes into the real buffer won't need to copy the data, just the Pieces.
 4. Now, nothing should be using runes.Buffer. Everything should be using runes.Buffer2. Remove runes.Buffer and rename runes.Buffer2 to runes.Buffer.
 5. If the new runes.Buffer isn't too big, consider removing the edit/runes package altogether and merging runes.Buffer into edit.Buffer.

**Comment by eaburns** [comment]
# Thoughts on caching

T's current buffer implementation is based around fixed-size blocks. When editing, the current block is cached in memory and edited solely in memory. This means that most edits only touch memory, not disk.

The piece table requires each ADD edit to write to the Add File, and each READ edit to read from the Add File. These are filesystem operations, which may make the piece table actually more expensive than the fixed-size block implementation that T currently uses. We need caching.

One simple way to get caching with the piece table is to implement a bounded-size, LRU cache where the cache elements are pieces. Here:

```
type piece struct {
	nbytes, nrunes int // Bounded by MaxBytes.
	offs int64 // Byte offset into Add File.
}
type buffer struct {
	pieces []piece
	// The cache.
	// Heap, ordered by least-recently-used first.
	// To look up the data for a piece,
	// linear search for a cacheBlock with the same offs
	// as the piece.
	cache []*cacheBlock
}
type cacheBlock struct {
	offs int64	// Byte offset into the Add File.
	used time.Time
	dirty bool
	data []byte // or []rune?
}

// Data returns the data for the piece.
// If the piece is in the cache:
//	The used time is updated and the cache is re-heapified,
// 	to maintain LRU order.
//
// If the piece is not in the cache:
// 	If the cache is too full to fit the piece's data:
// 		Remove LRU block from cache.
// 		If the block is dirty, write it.
// 		Repeat until there is room.
// 	Read the piece from the Add File into the cache.
func (b buffer) readPiece(p piece) ([]byte, error)

// AddPiece adds a piece to the buffer with the given data.
// The piece is not in the cache, since it's being added now.
// If the cache is too full to fit the piece's data:
// 	Remove LRU block from cache.
// 	If the block is dirty, write it.
// 	Repeat until there is room.
// Add the piece to the cache, marked as dirty.
func (b buffer) addPiece(p piece, d []byte) error
```

When a piece is deleted, no need to update the cache. The piece will be flushed lazily.

When splitting a piece, no need to update the cache. Splitting a piece is the same as deleting a piece and adding multiple pieces in its stead.

**Comment by eaburns** [comment]
I am skeptical that this is a good idea. Regexp addresses need rune-level indexing, which is inefficient with UTF8 storage. I looked into rewriting re1 using a RuneReader-style interface, but again, I determined that it won't work.

re1 needs access to the previous rune --- the one preceeding the first rune against which matching occurs --- in order to determine if the rune is a `\n`. This is required for matching `^`. For example, with the buffer
```
abcxyz
xyz
```
The address `#3+/^xyz/` should match the second occurrence of `xyz` (the one preceded by a newline) but not the first. For re1 to determine that the first occurrence of `xyz` is not a match it needs to look at the rune _preceeding_ the the first rune against which matching occurs, rune `#2,#3` to see whether it's `\n`.

**Comment by eaburns** [comment]
I suppose that the re1 interface could be changed to:
`func (re *Regexp) Match(rs io.RuneScanner, startOfLine bool) ([][2]int, error)`

**Comment by eaburns** [comment]
That doesn't make it any easier. Wrapping is hard.

**Comment by eaburns** [comment]
This is too hard. Let's just give up.

**Comment by eaburns** [comment]
Wait. What? Why does re1 need to do wrapped matching? I labored long to get that, but now it just seems wrong. We don't want regexp addresses to wrap. If they wrapped, `5+/abc/` could match an abc before line 5. That doesn't seem right.

**Comment by eaburns** [comment]
According to the sam man page, they do wrap around:
> The substring that matches the regular expression, found by looking toward the end (/) or beginning (?) of the file, and if necessary continuing the search from the other end to the starting point of the search.

**Comment by eaburns** [comment]
After #172, there will be no more worries about re1 blocking this. So this remains open.

**Comment by eaburns** [comment]
How will undo/redo on the piece table update mark addresses?

Typically their addresses are updated on each undone/redone change. If piece table undo/redo just restores the table state, how will we compute the changes in order to update the addresses?

**Comment by gedw99** [comment]
Crdt might help ,?

**Comment by eaburns** [comment]
Thanks for the suggestion. I'm not familiar with CRDTs, but according to Wikipedia they are useful for distributed algorithms. The plan here is to be completely centralized. Let me explain further and see if you still think they are appropriate. If so, could you please point me at some reading material?

The problem here is updating the address range of a mark during an undo. For example, if #2,#3 is marked as 'm and rune #0,#1 is deleted, 'm would update to #1,#2; a rune before it was deleted, so it moves one rune earlier. If we don't know how the buffer contents changed during an undo (the deltas) but only what the contents changed to, how can we update the address?

Here's a concrete example:

Before an undo, the buffer text is "abcdef" and mark 'm is the empty string after "abc" and before "def". After undo, the buffer becomes "a1234bcdef". Where should we move 'm? It's hard to say. Maybe it remains between the c and the d?

Now, what if I told you the delta: the undo changed "abcde" to "a1234bcde". Now we can follow the standard update rules. The update completely overlaps 'm, so 'm moves to the beginning of the change: in this case #0.

**Comment by gedw99** [comment]
I read the description, but i have to admit its a bit hard to follow what your system needs to do.
I posted because from all the notes of saw your struggling with, it looked like CRDT would be a different approach.
Also its Monday morning, and i have not had my coffee yet !!!

But CRDT is great where you have replicated data types. Its a bit like a quantum entangled particle, but with the delay. All potential operations will arrive at the same state, no mater the order.
So this allows you to keep state on both sides (client and server or whatever we want to call it) of the process boundary.
There areas that CRDT also touches is what is called Operational Transform. OT muates the events instead.
http://stackoverflow.com/questions/26694359/differences-between-ot-and-crdt

I was surprised to see about 6 golang implementation on github.
This one looks pretty sweet: https://github.com/tswindell/go-crdt

The thing about CRDT is that it only has a limited set of operations it can do. Its a bit like Redis compared to a SQL database, you could say. You only get the primitives.
But from the gist of what i saw you doing, it might work for you.
The other cool thing about CRTS, is that if the network connection drops or whatever, it will catch up, even if others have written to the same data from on the server. 
Google Docs and other collaborative apps using CRDT OR OT. I cant remember which

**Comment by eaburns** [comment]
Sorry if my description was a bit obtuse. It used sam parlance to describe addresses, and I wrote it quickly on my tablet, so it may be a bit dense.

It sounds like CRDTs solve a different problem, but a problem that I've also been thinking about a bit: how to ensure that clients show consistent state in the face of concurrent edits. The solution that I was thinking about was something really clumsy; every time the client requests state from the server, it will request *all* of the state (no partial updates). But now I'll checkout CRDTs. Thanks.

**Comment by gedw99** [comment]
Cool. Glad I could help.

**Comment by gedw99** [comment]
its really interesting & useful what your building to me btw. 

I was wondering if you can update the Readme to describe the thing at a higher level.
i might have time to contribute.

**Comment by eaburns** [comment]
Some day I will, but not quite yet.
Please feel free to email me. Let's keep the discussion here about issue #134.

**Comment by gedw99** [comment]
Sure.
